### ISSUE TITLE
Would it be interesting/worthwhile to investigate the recent spike increase in performance? Starting around d9a5ed48

### ISSUE THREAD

**Issue by wctgit** [gcp/leela-zero#751]
From d9a5ed48 to 6169f3fe we are seeing a steady succession of new networks that are improving one after another in a short period of time, and even as I write this a new network 5773f44c seems already to overtake 6169f3fe.

This makes me curious what the reason for the sudden spike is. Has the network started to learn a crucial part of the game of Go? Has the NN algorithm wandered from a flat region of the fitness landscape into a deep (or tall) region where certain combinations of weights lead to rapid improvements? Is this affecting the opening? End game? Reading ladders? Life and death situations? Better policy network or value network? Is the network beginning to perceive wider global game positional information, or is it becoming better at local reading of fights?

Although I'm curious about all this, I don't know how to go about investigating any of these questions. For one, I'm a bit of a noob at Go, and for two, I'm also a bit of a noob at NNs (I know the general concept well, but don't know the details or intricacies of it; e.g. the learning algorithm is currently beyond me).

Is anyone else interested/curious about this recent spike? Would it be worthwhile/interesting to spend some time investigating it, maybe to get some insight into a) the game of Go, b) the 'fitness landscape' the NN is currently exploring, c) the nature of NNs themselves, d) the performance of the current Leela Zero implementation (maybe there could be something for us to learn that would help improve training speed/efficiency), etc.? How could we go about doing such an investigation? Are there existing tools/functions that can be used to explore the recent networks? If not, would it be worth some time to try to develop such diagnostic tools?

It all seems pretty exciting to me. Am I perhaps getting excited over nothing very interesting? In other words, are these kinds of long plateaus with dramatic spikes a typical feature of NN training, such that this is all 'normal' and expected behaviour of NN training?

**Comment by gcp** [comment]
See #747. It's a fix in the training code.

**Comment by wctgit** [comment]
Briefly, could you summarize how the fix led to this dramatic improvement in training? I understand the concept of splitting data into training data vs. validation data, in a general AI/ML sense. How specifically does it affect the training algorithm of Leela Zero? What has the split accomplished that wasn't occurring properly before?

Sorry if these are naive questions. I'm very curious though! :-)

**Comment by gcp** [comment]
>https://github.com/gcp/leela-zero/pull/747#issuecomment-360694274

Basically, the TensorFlow shuffle buffer is by itself not enough to mix data from different games during training (unless it's made huge, which takes too much RAM, etc).

This was accidentally discovered because for the validation code I wanted to avoid having a second shuffle buffer so I enforced the mixing of games in another way.

**Comment by l1t1** [comment]
nice job

**Comment by wctgit** [comment]
Ah, and by training too much on the same games within some timeframe, the NN was getting overfitted to those specific games (and then later overfitted to the remaining games)? So more random shuffling allowed the learning algorithm to avoid falling into such local ruts, gaining a more generalized picture of the training data? Something like that?

BTW, what is the role of TensorFlow? Is it a package for tensor calculations, so we can emulate TPUs, I imagine? (Again, perhaps collecting this kind of information in the wiki could make a good resource for people curious about this project to get a good overview of the project. I am aware of the FAQ. Some of the FAQ info would fit well in a wiki, too. The wiki could provide a more comprehensive reference than the FAQ, IMHO.)

**Comment by gcp** [comment]
>Something like that?

Yes, that's a good summary.

>BTW, what is the role of TensorFlow? 

https://en.wikipedia.org/wiki/TensorFlow

It's together with PyTorch one of the most popular frameworks for Deep Learning. (That's the only reason I chose it - I was using Caffe before and thought we'd get more contributors with a more modern/popular framework)

**Comment by wctgit** [comment]
Wow, cool. BTW, I have some experience with contributing to wikis. I was one of the later contributors to the original wiki, Ward Cunningham's WikiWikiWeb. I may not be able to contribute to the coding, due to burn out, basically, but I would be more than happy to contribute to gathering info into a wiki. Is it possible? I am pretty much a noob at GitHub, too. (^_^;)

**Comment by wctgit** [comment]
Actually, I wonder if you might consider re-opening this issue: It still might be worthwhile investigating what kinds of improvements to the NN have occurred due to the improvement in the de-correlation of training data. It might reveal information that could lead to even greater improvements to the training algorithms. Just a thought.
